### PR TITLE
Send an email from Go using test4impact account

### PR DESCRIPTION
**Pull request by yoninachmany**
-Using "net/smtp" package, send email that uses os environments for email and password

### CODE REVIEW

File: transcription/email.go [hunk 1]
```diff
@@ -0,0 +1,30 @@
...
+func sendEmail(from string, to []string, subject string, body string) {
+	// Set up authentication information.
+	auth := smtp.PlainAuth(
+		"",
+		os.Getenv("MAIL_EMAIL"),
+		os.Getenv("MAIL_PASSWORD"),
+		"smtp.gmail.com",
+	)
+
+	// Connect to the server, authenticate, set the sender and recipient,
```
Comment (sandlerben) [inline transcription/email.go]: These comments aren't necessary. I think the code is quite straightforward.

File: transcription/email.go [hunk 2]
```diff
@@ -0,0 +1,30 @@
+package transcription
+
+import (
+	"log"
+	"net/smtp"
+	"os"
+	"strings"
+)
+
+func sendEmail(from string, to []string, subject string, body string) {
```
Comment (sandlerben) [inline transcription/email.go]: Do you think this function will be used outside the transcription package?

File: transcription/email.go [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+	// Set up authentication information.
+	auth := smtp.PlainAuth(
+		"",
+		os.Getenv("MAIL_EMAIL"),
+		os.Getenv("MAIL_PASSWORD"),
+		"smtp.gmail.com",
+	)
```
Comment (sandlerben) [inline transcription/email.go]: There is no character limit convention in Go. How come you don't want to put this all on the same line? Do you think this is more readable?

File: transcription/email.go [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+
+	// Connect to the server, authenticate, set the sender and recipient,
+	// and send the email all in one step.
+	msg := []byte("To: " + strings.Join(to, ", ") + "\r\n" +
```
Comment (sandlerben) [inline transcription/email.go]: Why do you need to define `msg`? Shouldn't the email message be passed as an argument?

File: transcription/email.go [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+		"Subject: " + subject + "\r\n" +
+		"\r\n" +
+		body + "\r\n")
+	err := smtp.SendMail("smtp.gmail.com:25", auth, from, to, msg)
+
+	if err != nil {
```
Comment (sandlerben) [inline transcription/email.go]: The way you have it now is fine, but I think the more idiomatic Go way is like

``` go
if err := someFunction(); err != nil {
    // do stuff
}
```

File: transcription/email.go [hunk 6]
```diff
@@ -0,0 +1,30 @@
...
+	)
+
+	// Connect to the server, authenticate, set the sender and recipient,
+	// and send the email all in one step.
+	msg := []byte("To: " + strings.Join(to, ", ") + "\r\n" +
+		"Subject: " + subject + "\r\n" +
+		"\r\n" +
+		body + "\r\n")
+	err := smtp.SendMail("smtp.gmail.com:25", auth, from, to, msg)
+
```
Comment (sandlerben) [inline transcription/email.go]: no space here

File: transcription/email.go [hunk 7]
```diff
@@ -0,0 +1,30 @@
...
+		"smtp.gmail.com",
+	)
+
+	// Connect to the server, authenticate, set the sender and recipient,
+	// and send the email all in one step.
+	msg := []byte("To: " + strings.Join(to, ", ") + "\r\n" +
+		"Subject: " + subject + "\r\n" +
+		"\r\n" +
+		body + "\r\n")
+	err := smtp.SendMail("smtp.gmail.com:25", auth, from, to, msg)
```
Comment (sandlerben) [inline transcription/email.go]: Should `"smtp.gmail.com:25"` be a string literal? It seems like something which should be passed in as an argument to this function.

Comment (sandlerben) [inline transcription/email.go]: For now, you don't have to worry about making it an argument, but I would assign it to its own variable

``` go
somethingSomething := "smtp.gmail.com:25"
```

File: transcription/email.go [hunk 8]
```diff
@@ -0,0 +1,30 @@
...
+
+	if err != nil {
+		log.Fatal(err)
```
Comment (sandlerben) [inline transcription/email.go]: We should talk at some point about how error handling is done in Go. The best thing to do here would be to return err so the **caller** has to deal with it.

One of my favorite things about Go is that errors are always handled. The key reason why is that error values are propagated up the call stack so the **right** code handles it. Imagine we call `sendEmail` from another function `foo`. If `sendEmail` has an error, then `foo` might know how to handle it. Or maybe `foo` just panics. Or maybe `foo` was called by `bar` so `foo` returns the error back to bar. Regardless, `sendEmail` should have **no say** in how this error is handled. It should just say "not my problem" and move on.

File: transcription/email.go [hunk 9]
```diff
@@ -0,0 +1,30 @@
+package transcription
```
Comment (sandlerben) [inline transcription/email.go]: Let's just put this all in one file `transcription.go` for now.

**Comment by sandlerben** [comment]
Try to write better commit messages https://www.dropbox.com/s/vgj61cvaeto41gg/Screenshot%202016-02-25%2001.50.30.png?dl=0

https://github.com/erlang/otp/wiki/Writing-good-commit-messages

File: transcription/email.go (hunk 8 above)
Comment (yoninachmany) [inline transcription/email.go]: for reference, took most code from `smtp` example code for `SendEmail`
screen shot 2016-02-25 at 9 34 02 am

**Comment by yoninachmany** [comment]
@sandlerben how does this look?

File: transcription/transcription.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+	toHeader := "To: " + strings.Join(to, ", ")
+	subjectHeader := "Subject: " + subject
+	msgHeaders := []string{fromHeader, toHeader, subjectHeader}
+
+	// The msg parameter should be an RFC 822-style email with headers first,
+	// a blank line, and then the message body. The lines of msg should be CRLF terminated.
+	msg := []byte(strings.Join(msgHeaders, "\r\n") +
+		"\r\n" +
+		body + "\r\n")
+	addr := host + ":" + string(port)
```
Comment (sandlerben) [inline transcription/transcription.go:18]: Just make one argument `host` which includes the port.

File: transcription/transcription.go [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+	"net/smtp"
+	"strings"
+)
+
+// SendEmail authenticates, creates message, and sends email from arguments.
+func SendEmail(username string, password string, host string, port int, to []string, subject string, body string) error {
+	auth := smtp.PlainAuth("", username, password, host)
+
+	// The msg headers should usually include fields such as "From", "To", "Subject", and "Cc".
+	// Sending "Bcc" messages is accomplished by including an email address in
```
Comment (sandlerben) [inline transcription/transcription.go]: Do we need to document this here?

File: transcription/transcription.go [hunk 12]
```diff
@@ -0,0 +1,31 @@
...
+	// the to parameter but not including it in the msg headers.
+	from := username
+	fromHeader := "From: " + from
+	toHeader := "To: " + strings.Join(to, ", ")
+	subjectHeader := "Subject: " + subject
+	msgHeaders := []string{fromHeader, toHeader, subjectHeader}
```
Comment (sandlerben) [inline transcription/transcription.go]: Can you pull this logic out and put it in its own function?

File: transcription/transcription.go (hunk 10 above)
Comment (sandlerben) [inline transcription/transcription.go:18]: Oh wait it looks like you sometimes need the host without the port. This is fine.

File: transcription/transcription.go [hunk 13]
```diff
@@ -0,0 +1,31 @@
+package transcription
+
+import (
+	"net/smtp"
+	"strings"
+)
+
+// SendEmail authenticates, creates message, and sends email from arguments.
+func SendEmail(username string, password string, host string, port int, to []string, subject string, body string) error {
```
Comment (sandlerben) [inline transcription/transcription.go:11]: Can we document what some of these arguments mean? I don't think it's self-explanatory what `host` or `port` mean.

File: transcription/transcription.go (hunk 12 above)
Comment (sandlerben) [inline transcription/transcription.go]: The function can return `msgHeaders` which is an already-joined string.

File: transcription/transcription.go [hunk 14]
```diff
@@ -0,0 +1,31 @@
...
+	// the to parameter but not including it in the msg headers.
+	from := username
+	fromHeader := "From: " + from
+	toHeader := "To: " + strings.Join(to, ", ")
+	subjectHeader := "Subject: " + subject
+	msgHeaders := []string{fromHeader, toHeader, subjectHeader}
+
+	// The msg parameter should be an RFC 822-style email with headers first,
+	// a blank line, and then the message body. The lines of msg should be CRLF terminated.
+	msg := []byte(strings.Join(msgHeaders, "\r\n") +
```
Comment (sandlerben) [inline transcription/transcription.go]: Is msg the best name for this variable?

File: transcription/transcription.go (hunk 11 above)
Comment (yoninachmany) [inline transcription/transcription.go]: Removing BCC line

File: transcription/transcription.go (hunk 12 above)
Comment (yoninachmany) [inline transcription/transcription.go]: gonna have it return the `[]string` cuz i want to keep all the CRLF termination characters in the same function

Comment (yoninachmany) [inline transcription/transcription.go]: jk looks cleaner when string returned

File: transcription/transcription.go (hunk 14 above)
Comment (yoninachmany) [inline transcription/transcription.go]: From `smtp` library

screen shot 2016-02-25 at 6 07 12 pm
screen shot 2016-02-25 at 6 07 01 pm

**Comment by yoninachmany** [comment]
@sandlerben addressed ur latest comments

File: transcription/transcription.go [hunk 15]
```diff
@@ -0,0 +1,31 @@
+package transcription
+
+import (
+	"net/smtp"
+	"strings"
+)
+
+// SendEmail authenticates, creates message, and sends email using arguments.
```
Comment (sandlerben) [inline transcription/transcription.go]: Can we fix up the wording of this comment. Authenticates what? Creates message -> create messages or creates a message. Using arguments is probably unnecessary.

File: transcription/transcription.go [hunk 16]
```diff
@@ -0,0 +1,31 @@
...
+	// a blank line, and then the message body. The lines of msg should be CRLF terminated.
+	msg := []byte(msgHeaders(from, to, subject) + "\r\n" + body + "\r\n")
+	addr := host + ":" + string(port)
+	if err := smtp.SendMail(addr, auth, from, to, msg); err != nil {
+		return err
+	}
+	return nil
+}
+
+// The msg headers should usually include fields such as "From", "To", "Subject", and "Cc".
```
Comment (sandlerben) [inline transcription/transcription.go]: function comments should always starting with the name of the function.

How about "msgHeaders includes fields such as..."?

Also, is this comment necessary? Would someone reading the code not understand what msgHeaders does?

Comment (sandlerben) [inline transcription/transcription.go]: If you think it makes the code more readable, feel free to leave it.

**Comment by sandlerben** [comment]
@yoninachmany looks good to me

Let's get one more set of eyes on this. @roornelas can you give this a quick look?

**Comment by roornelas** [comment]
Looks good to me. merge?

**Comment by yoninachmany** [comment]
@sandlerben heads up to remake https://github.com/hack4impact/audio-transcription-service/pull/16 merging into master

**Pull request merged**